src/hostnet/lib/dns_forward.ml

```diff
@@ -26,95 +26,34 @@ let tidstr_of_dns dns =
-  mutable used_resolvers : (Ipaddr.t * int) list;
-  mutable last_use: float;
-}
-
-let table = Hashtbl.create 7
-
-let start_reaper () =
-  let rec loop () =
-    Time.sleep 60.
-    >>= fun () ->
-    let snapshot = Hashtbl.copy table in
-    let now = Unix.gettimeofday () in
-    Hashtbl.iter (fun k trec ->
-      if now -. trec.last_use > 60. then begin
-        Log.debug (fun f -> f "DNS[%04x] Expiring DNS trec" k);
-        Hashtbl.remove table k
-      end
-    ) snapshot;
-    loop () in
-  loop ()
-
-let input ~ip ~udp ~src ~dst ~src_port buf =
+let input ~secondary ~ip ~udp ~src ~dst ~src_port buf =
   if List.mem dst (Ip.get_ip ip) then begin
 
   let src_str = Ipaddr.V4.to_string src in
   let dst_str = Ipaddr.V4.to_string dst in
 
   let dns = parse_dns buf in
 
-  let remove_tid dns =
-    match dns with
-    | (_, None) -> ()
-    | (_, Some { Dns.Packet.id; _ }) -> Hashtbl.remove table id in
-
   Log.debug (fun f -> f "DNS[%s] %s:%d -> %s %s" (tidstr_of_dns dns) src_str src_port dst_str (string_of_dns dns));
 
-  match dns with
-  | (_, None) -> begin
-    Resolv_conf.get ()
-    >>= function
-    | r::_ -> Lwt.return_some r
-    | _ -> Lwt.return_none
-  end
-  | (_, Some { Dns.Packet.id = tid; _ }) -> begin
-    Lwt.catch
-      (fun () ->
-        let trec = Hashtbl.find table tid in
-        let r, rs, urs = match trec.resolvers with
-          | r::rs -> (r,rs,r::trec.used_resolvers)
-          | _ -> match List.rev trec.used_resolvers with
-            | r::rs -> (r,rs,[])
-            | _  -> assert false (* resolvers and used_resolvers cannot both be empty *)
-        in
-        Log.debug (fun f -> f "DNS[%s] Retry" (tidstr_of_dns dns));
-        trec.resolvers <- rs;
-        trec.used_resolvers <- urs;
-        trec.last_use <- Unix.gettimeofday ();
-        Lwt.return_some r
-      )
-      (function
-      | Not_found -> begin
-         (* Re-read /etc/resolv.conf on every request. This ensures that
-            changes to DNS on sleep/resume or switching networks are reflected
-            immediately. The file is very small, and parsing it shouldn't be
-            too slow. *)
-        Resolv_conf.get ()
-        >>= function
-        | r::rs -> begin
-          let last_use = Unix.gettimeofday () in
-          let trec = { resolvers = rs; used_resolvers = r :: []; last_use } in
-          Hashtbl.replace table tid trec;
-          Lwt.return_some r
-        end
-        | _ -> Lwt.return_none
-      end
-      | ex -> Lwt.fail ex
-      )
-  end;
+  Resolv_conf.get ()
+
+  >>= ( function
+  | _::sec::_ when secondary -> Lwt.return_some ("secondary", sec)
+  | pri::_::_ when not secondary -> Lwt.return_some ("primary", pri)
+  (* These two could be collapsed, but for the desire to differentiate in the logging *)
+  | pri::_ when secondary -> Lwt.return_some ("secondary(wrapped)", pri)
+  | pri::_ when not secondary -> Lwt.return_some ("primary(sole)", pri)
+  | _ -> Lwt.return_none )
```

Comment:
Without them the final `| _ -> Lwt.return_none` was also capturing the following bind, leading to type hilarity. @djs55 and @avsm advised that the parens were the normal way to solve that.